text/0000-ide.md

```diff
@@ -0,0 +1,530 @@
+mechanism to the compiler to support different formats for error messages. We
+already structure our error messages to some extent (separating the span
+information, the message, and the error code). Rather than turning these
+components into text in a fairly ad hoc manner, we should preserve that
+structure, and some central error handler should convert into a chosen format.
+We should support the current text format, JSON (or some other structured
+format) for tools to use, and HTML for rich error messages (this is somewhat
+orthogonal to this RFC, but has been discussed in the past as a desirable
+feature).
+
+
+# Drawbacks
+
+It's a lot of work. On the other hand the largest changes are desirable for
+general improvements in compilation speed or for other tools.
+
+
+# Alternatives
+
+The oracle and quick-check compiler could be combined in a single tool. This
+might be more efficient, but would increase complexity and decrease opportunity
+for third party alternatives.
+
+The oracle could do more - actually perform some of the processing tasks usually
+done by IDEs (such as editing source code) or other tools (refactoring,
+reformating, etc.).
+
+Should the oracle hide the quick-check compiler? I.e., the IDE talks only to the
+oracle and the oracle requests compilation as needed. This might make things a
+bit simpler for the IDE and means less IPC overhead and complexity. Either the
+oracle could be responsible for all coordination, or the IDE could remain
+responsible for coordinating when crates are handled, and the oracle is
+responsible for coordinating calls to the quick check compiler to build a single
+crate.
+
+
+# Unresolved questions
+
+Should the quick-check compilation be provided by a separate tool or a mode of
+the compiler? It is fairly different in its operation from the compiler. It
+might be better to provide a different 'frontend' rather than adding many more
+options to the compiler. (I think the answer is 'yes').
+
+Should quick-check be a long running process? It could save some time by not
+having to reload metadata, but having to keep metadata for an entire project in
+memory would be expensive. We could perhaps compromise by unloading when the
+user needs to recompile a different crate. I believe it is probably better in
+the long run, but a batch process is OK to start with.
+
+How and when should we generate crate metadata. It seems sensible to generate
+this when we switch to editing/re-compiling a different crate. However, it's not
+clear if this must be done from scratch or if it can be produced from the
+incremental compilation metadata (see that RFC, I guess).
+
+What should we call the oracle tool? I don't particularly like "oracle",
+although it is descriptive (it comes from the Go tool of the same name).
+Alternatives are 'Rider', 'Racer Server', or anything you can think of.
+
+How do we handle different versions of Rust and interact with multi-rust?
+Upgrades to the next stable version of Rust?
```

Comment:
I too was wondering how to deal with multiple versions of Rust. In my IDE plugin, I would like to let the user choose its compiler, and be able to invoke the nightly compiler or the stable one explicitly using different targets. I don't know how I could give this choice to the user, for the code model used for navigation/completion/refactoring.